internal/nearcache/repairing_task.go

```diff
@@ -0,0 +1,568 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nearcache
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pubcluster "github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/invocation"
+	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto/codec"
+	"github.com/hazelcast/hazelcast-go-client/internal/serialization"
+	"github.com/hazelcast/hazelcast-go-client/types"
+)
+
+/*
+ReparingTask runs on Near Cache side and only one instance is created per data-structure type like IMap and ICache.
+Repairing responsibilities of this task are:
+
+    * To scan RepairingHandlers to see if any Near Cache needs to be invalidated according to missed invalidation counts (controlled via MaxToleratedMissCount).
+    * To send periodic generic-operations to cluster members in order to fetch latest partition sequences and UUIDs (controlled via InvalidationMinReconciliationIntervalSeconds.
+
+See: com.hazelcast.internal.nearcache.impl.invalidation.RepairingTask
+*/
+type ReparingTask struct {
+	reconciliationIntervalNanos int64
+	lastAntiEntropyRunNanos     int64
+	handlers                    *sync.Map
+	ss                          *serialization.Service
+	ps                          *cluster.PartitionService
+	doneCh                      <-chan struct{}
+	invalidationMetaDataFetcher InvalidationMetaDataFetcher
+	lg                          ilogger.LogAdaptor
+	maxToleratedMissCount       int
+	partitionCount              int32
+	running                     int32
+}
+
+func NewReparingTask(recInt int, maxMissCnt int, ss *serialization.Service, ps *cluster.PartitionService, lg ilogger.LogAdaptor, mf InvalidationMetaDataFetcher, doneCh <-chan struct{}) *ReparingTask {
+	nc := &ReparingTask{
+		reconciliationIntervalNanos: int64(recInt) * 1_000_000_000,
+		maxToleratedMissCount:       maxMissCnt,
+		invalidationMetaDataFetcher: mf,
+		doneCh:                      doneCh,
+		ss:                          ss,
+		ps:                          ps,
+		lg:                          lg,
+		handlers:                    &sync.Map{},
+		partitionCount:              ps.PartitionCount(),
+	}
+	return nc
+}
+
+func (rt *ReparingTask) start() {
+	// see: com.hazelcast.internal.nearcache.impl.invalidation.RepairingTask#scheduleNextRun
```

Comment:
I think `scheduleNextRun` schedules only one shot run not infinite runs like you did.

Reply:
> Java client re-schedules after the handler is run, almost the same thing. See: com.hazelcast.spi.impl.executionservice.TaskScheduler#schedule(java.lang.Runnable, long, java.util.concurrent.TimeUnit) which schedules run after one second which in turn calls scheduleNextRun again.

yeah I noticed that you schedule a bit different than Java. I am worried if there is a behavior difference. I did not notice but there may be something I missed. 

> Also we do not have this protection, with saying that I have no clue what it serves https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/invalidation/RepairingTask.java#L130

Seems like running always true after it is set to true, so the check is always true.

Reply:
I changed this bit at: https://github.com/hazelcast/hazelcast-go-client/pull/807/commits/c67bb647844dec1c173a226819e5f398deac8fe2

Reply:
@utku-caglayan to resolve since he asked for ticker reset, not me.